src/main.rs

```diff
@@ -295,113 +299,95 @@ fn main() {
-                .expect("None stream errors");
-
-            match next {
-                Either::A((Either::A((signal, _)), _)) => {
-                    let signal = signal.expect("Signal streams never end");
-                    if signal == SIGHUP {
-                        reload_config(&opts, &mut topology, &mut rt)
-                    } else {
-                        break signal;
-                    }
-                }
-                Either::A((Either::B((_crash, _)), _)) => break SIGINT, // Trigger graceful shutdown if a component crashed
-                Either::B((_, _)) => reload_config(&opts, &mut topology, &mut rt),
-            }
+        // Arc and Mutex are here to satisfy 'static/Send constraints imposed by block_on method.
+        let signals = std::sync::Arc::new(std::sync::Mutex::new(MultiplexStream::new()));
+        let signals_clone = signals.clone();
+        // Constructs futures that poll signal Stream, also exists because of block_on.
+        let signal_future = move || {
+            let signals = signals_clone.clone();
+            future::poll_fn(move || signals.lock().expect("Not poisoned").poll())
         };
+        signals
+            .lock()
+            .expect("Not poisoned")
+            .add_as(
+                graceful_crash.map_err::<String, _>(|()| unreachable!()),
+                SignalEvent::Shutdown,
+            )
+            .add_as(
+                config_watcher.map_err::<String, _>(|()| unreachable!()),
+                SignalEvent::ReloadConfig,
+            );
 
-        if signal == SIGINT || signal == SIGTERM {
-            info!("Shutting down.");
-            let shutdown = topology.stop();
-            metrics_trigger.cancel();
+        #[cfg(unix)]
+        {
+            signals
+                .lock()
+                .expect("Not poisoned")
+                .add_as(Signal::new(SIGINT).flatten_stream(), SignalEvent::Shutdown)
+                .add_as(Signal::new(SIGTERM).flatten_stream(), SignalEvent::Shutdown)
+                .add_as(Signal::new(SIGQUIT).flatten_stream(), SignalEvent::Quit)
+                .add_as(
+                    Signal::new(SIGHUP).flatten_stream(),
+                    SignalEvent::ReloadConfig,
+                );
+        }
+        #[cfg(windows)]
+        {
+            signals.lock().expect("Not poisoned").add_as(
+                tokio_signal::ctrl_c().flatten_stream(),
+                SignalEvent::Shutdown,
+            );
+        }
 
-            match rt.block_on(shutdown.select2(signals.into_future())) {
-                Ok(Either::A(_)) => { /* Graceful shutdown finished */ }
-                Ok(Either::B(_)) => {
-                    info!("Shutting down immediately.");
-                    // Dropping the shutdown future will immediately shut the server down
+        // Process signals in running state
+        loop {
+            let signals = signals.clone();
+            match rt.block_on(signal_future()) {
```

Comment:
Instead of using poll_fn here I'd like to use this combinator https://docs.rs/futures/0.1.29/futures/stream/struct.StreamFuture.html

Our signal module should expose a Stream of signals, it shouldn't matter where they come from but we can then use this combinator with block_on.

Reply:
That would be much better. It will also remove the need for Arc<Mutext<_>>. 
Great find.

Reply:
Actually I forgot about it.